acvm/src/lib.rs

```diff
@@ -92,7 +97,26 @@ fn first_missing_assignment(
     })
 }
 
-pub trait Backend: SmartContract + ProofSystemCompiler + PartialWitnessGenerator + Default {}
+pub trait Backend:
+    SmartContract + ProofSystemCompiler + PartialWitnessGenerator + ReferenceString + Default
+{
+}
+
+#[async_trait]
+pub trait ReferenceString {
+    /// The Error type returned by failed function calls in the ReferenceString trait.
+    type Error: std::error::Error; // fully-qualified named because thiserror is `use`d at the top of the crate
+
+    /// Provides the reference string that is needed by other traits
+    async fn get_reference_string(&self, circuit: &Circuit) -> Result<Vec<u8>, Self::Error>;
+
+    /// Validates a (likely cached) reference string against a circuit
+    fn is_reference_string_valid(
```

Comment:
what is this supposed to do?

Reply:
How can a backend validate this if not generating it again and comparing with the input?

Reply:
The validity of the download should be done when it is downloaded, no need extra interface for this.
File corruption should be handled by the one caching it (in that case it is cargo), no need extra interface for this.
The circuit size test makes sense but it is dependent on the proving system so it is not generic. I am not sure if this works well for a groth16 backend?

Reply:
A cached file can be corrupted by the filesystem, so you can't just do it once and forget about it.

Reply:
sure, but you can hash it and checks if it is corrupted or not.

Reply:
Who is hashing it? What are you comparing it to?

Reply:
you hash it when you retrieve it from the backend, and you compare this hash from the hash of the file

Reply:
This API has been changed to be `update_common_reference_string` to account for downloading only the additional CRS data that is needed (instead of the entirety of the data). It is worth noting that Nargo CLI will need to write data in every scenario with this API but it will lighten network loads, so I think it is a fine trade-off.